File: test/showcase/scala/src/main/scala/com/gekka/showcase/DdataActors.scala
package com.gekka.showcase

import org.apache.pekko.actor.{Actor, ActorLogging, Cancellable, Props}
import org.apache.pekko.cluster.Cluster
import org.apache.pekko.cluster.ddata.Replicator._
import org.apache.pekko.cluster.ddata._

import scala.concurrent.duration._
import scala.util.Random

object DdataActors {
  def all(selfLabel: String): List[Props] = List(
    GCounterActor.props(selfLabel),
    PNCounterActor.props(selfLabel),
    GSetActor.props(selfLabel),
    ORSetActor.props(selfLabel),
    ORMapActor.props(selfLabel),
    ORFlagActor.props(selfLabel),
    LWWRegisterActor.props(selfLabel),
    LWWMapActor.props(selfLabel),
  )

  val names: List[String] = List(
    "ddata/gcounter", "ddata/pncounter", "ddata/gset", "ddata/orset",
    "ddata/ormap", "ddata/orflag", "ddata/lwwregister", "ddata/lwwmap",
  )

  case object Tick

  abstract class CrdtActor[T <: ReplicatedData](selfLabel: String, keyName: String) extends Actor with ActorLogging {
    import context.dispatcher
    protected val replicator = DistributedData(context.system).replicator
    protected implicit val node: SelfUniqueAddress = DistributedData(context.system).selfUniqueAddress
    protected val rng = new Random()
    private var tick: Cancellable = _

    protected def tickMs: Long = 1000L + rng.nextInt(4000) // [1s, 5s)

    override def preStart(): Unit = {
      tick = context.system.scheduler.scheduleOnce(tickMs.millis, self, Tick)
    }

    override def postStop(): Unit = Option(tick).foreach(_.cancel())

    protected def reschedule(): Unit = {
      tick = context.system.scheduler.scheduleOnce(tickMs.millis, self, Tick)
    }

    override def receive: Receive = handle.orElse {
      case _: UpdateResponse[_] => ()
      case _: GetResponse[_]    => ()
      case other =>
        log.error(s"${getClass.getSimpleName}: unexpected: ${other.getClass.getName}")
    }

    def handle: Receive
  }
}

object GCounterActor {
  def props(selfLabel: String): Props = Props(new GCounterActor(selfLabel))
}
class GCounterActor(selfLabel: String) extends DdataActors.CrdtActor[GCounter](selfLabel, "g-counter") {
  private val key = GCounterKey("showcase-gcounter")
  override def handle: Receive = {
    case DdataActors.Tick =>
      if (rng.nextBoolean()) replicator ! Update(key, GCounter.empty, WriteLocal)(_ :+ 1)
      else                   replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

object PNCounterActor {
  def props(selfLabel: String): Props = Props(new PNCounterActor(selfLabel))
}
class PNCounterActor(selfLabel: String) extends DdataActors.CrdtActor[PNCounter](selfLabel, "pn-counter") {
  private val key = PNCounterKey("showcase-pncounter")
  override def handle: Receive = {
    case DdataActors.Tick =>
      val op = rng.nextInt(3)
      if (op == 0)      replicator ! Update(key, PNCounter.empty, WriteLocal)(_ :+ 1)
      else if (op == 1) replicator ! Update(key, PNCounter.empty, WriteLocal)(_ decrement 1)
      else              replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

object GSetActor {
  def props(selfLabel: String): Props = Props(new GSetActor(selfLabel))
}
class GSetActor(selfLabel: String) extends DdataActors.CrdtActor[GSet[String]](selfLabel, "g-set") {
  private val key = GSetKey[String]("showcase-gset")
  override def handle: Receive = {
    case DdataActors.Tick =>
      if (rng.nextBoolean()) replicator ! Update(key, GSet.empty[String], WriteLocal)(_ + s"item-${rng.nextInt(50)}")
      else                   replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

object ORSetActor {
  def props(selfLabel: String): Props = Props(new ORSetActor(selfLabel))
}
class ORSetActor(selfLabel: String) extends DdataActors.CrdtActor[ORSet[String]](selfLabel, "or-set") {
  private val key = ORSetKey[String]("showcase-orset")
  override def handle: Receive = {
    case DdataActors.Tick =>
      val op = rng.nextInt(3)
      if (op == 0)      replicator ! Update(key, ORSet.empty[String], WriteLocal)(_ :+ s"item-${rng.nextInt(50)}")
      else if (op == 1) replicator ! Update(key, ORSet.empty[String], WriteLocal)(_ remove s"item-${rng.nextInt(50)}")
      else              replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

object ORMapActor {
  def props(selfLabel: String): Props = Props(new ORMapActor(selfLabel))
}
class ORMapActor(selfLabel: String) extends DdataActors.CrdtActor[ORMap[String, ORSet[String]]](selfLabel, "or-map") {
  private val key = ORMapKey[String, ORSet[String]]("showcase-ormap")
  override def handle: Receive = {
    case DdataActors.Tick =>
      val k = s"k${rng.nextInt(10)}"
      if (rng.nextBoolean()) {
        replicator ! Update(key, ORMap.empty[String, ORSet[String]], WriteLocal) { m =>
          m.updated(node, k, ORSet.empty[String])(_ :+ s"v${rng.nextInt(100)}")
        }
      } else {
        replicator ! Get(key, ReadLocal)
      }
      reschedule()
  }
}

object ORFlagActor {
  def props(selfLabel: String): Props = Props(new ORFlagActor(selfLabel))
}
class ORFlagActor(selfLabel: String) extends DdataActors.CrdtActor[Flag](selfLabel, "or-flag") {
  private val key = FlagKey("showcase-orflag")
  override def handle: Receive = {
    case DdataActors.Tick =>
      if (rng.nextBoolean()) replicator ! Update(key, Flag.empty, WriteLocal)(_.switchOn)
      else                   replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

object LWWRegisterActor {
  def props(selfLabel: String): Props = Props(new LWWRegisterActor(selfLabel))
}
class LWWRegisterActor(selfLabel: String) extends DdataActors.CrdtActor[LWWRegister[String]](selfLabel, "lww-register") {
  private val key = LWWRegisterKey[String]("showcase-lwwregister")
  override def handle: Receive = {
    case DdataActors.Tick =>
      if (rng.nextBoolean())
        replicator ! Update(key, LWWRegister(node, s"init-$selfLabel"), WriteLocal)(_.withValueOf(s"v${rng.nextInt(1000)}"))
      else
        replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

object LWWMapActor {
  def props(selfLabel: String): Props = Props(new LWWMapActor(selfLabel))
}
class LWWMapActor(selfLabel: String) extends DdataActors.CrdtActor[LWWMap[String, String]](selfLabel, "lww-map") {
  private val key = LWWMapKey[String, String]("showcase-lwwmap")
  override def handle: Receive = {
    case DdataActors.Tick =>
      val k = s"k${rng.nextInt(10)}"
      if (rng.nextBoolean()) replicator ! Update(key, LWWMap.empty[String, String], WriteLocal)(_ :+ (k -> s"v${rng.nextInt(1000)}"))
      else                   replicator ! Get(key, ReadLocal)
      reschedule()
  }
}

File: cmd/showcase-gekka/ddata_actors.go
// cmd/showcase-gekka/ddata_actors.go — Gekka-side DDDT actors for FT3
// of the gekka_showcase_test plan.
//
// startDdata launches one goroutine per supported CRDT type. Each goroutine
// performs a random write or read against gekka's local Replicator every
// [1s, 5s). State propagates to peers via the Replicator's periodic gossip
// loop (already started by gekka.NewCluster), so no explicit fan-out is
// needed here.
//
// NOTE: gekka's built-in *ddata.Replicator currently exposes gossip wiring
// for SIX of Pekko's eight standard CRDTs:
//
//	GCounter, PNCounter, ORSet, ORFlag, LWWRegister, LWWMap.
//
// GSet and ORMap exist as CRDT types under cluster/ddata/ but the Replicator
// has no accessor or gossip path for them. Scala's Task 8 still exercises
// all 8 CRDTs Scala<->Scala; the cross-language FT3 convergence contract is
// therefore limited to the 6 CRDTs above on the gekka<->Scala leg.
//
// SPDX-License-Identifier: MIT
package main

import (
	"math/rand"
	"strconv"
	"time"

	gekka "github.com/sopranoworks/gekka"
	"github.com/sopranoworks/gekka/cluster/ddata"
	"github.com/sopranoworks/gekka/logger"
)

// Showcase CRDT keys. These must match the Scala-side key names in
// test/showcase/scala/src/main/scala/com/gekka/showcase/DdataActors.scala
// for the cross-language convergence smoke test (FT3).
const (
	keyGCounter    = "showcase-gcounter"
	keyPNCounter   = "showcase-pncounter"
	keyORSet       = "showcase-orset"
	keyORFlag      = "showcase-orflag"
	keyLWWRegister = "showcase-lwwregister"
	keyLWWMap      = "showcase-lwwmap"
)

// ddataTickInterval returns a pseudo-random tick duration in [1s, 5s).
// Mirrors the Scala-side CrdtActor scheduling jitter so cross-language load
// patterns are comparable.
func ddataTickInterval(rng *rand.Rand) time.Duration {
	return time.Duration(1000+rng.Intn(4000)) * time.Millisecond
}

// startDdata launches one goroutine per supported CRDT. The goroutines are
// fire-and-forget (no graceful shutdown wiring) — they exit when the process
// exits. Each goroutine owns its own *rand.Rand so the RNGs do not contend.
//
// startDdata runs on every node regardless of peer configuration. The local
// Replicator gossip loop discovers peers as the cluster membership reaches
// Up, and unicast updates the moment a peer becomes reachable.
func startDdata(cluster *gekka.Cluster, selfLabel string) {
	repl := cluster.Replicator()
	if repl == nil {
		// Defensive: NewCluster always wires a Replicator, but if a future
		// refactor makes it optional we don't want a nil-deref to crash the
		// whole showcase binary.
		logger.Default().Error("DdataActor: cluster.Replicator() returned nil; FT3 disabled")
		return
	}

	logger.Default().Info("DdataActor: starting CRDT loops",
		"selfLabel", selfLabel,
		"supported", []string{"gcounter", "pncounter", "orset", "orflag", "lwwregister", "lwwmap"},
		"unsupported", []string{"gset", "ormap"},
	)

	seed := time.Now().UnixNano()
	go gcounterLoop(repl, selfLabel, rand.New(rand.NewSource(seed+1)))
	go pncounterLoop(repl, selfLabel, rand.New(rand.NewSource(seed+2)))
	go orsetLoop(repl, selfLabel, rand.New(rand.NewSource(seed+3)))
	go orflagLoop(repl, selfLabel, rand.New(rand.NewSource(seed+4)))
	go lwwregisterLoop(repl, selfLabel, rand.New(rand.NewSource(seed+5)))
	go lwwmapLoop(repl, selfLabel, rand.New(rand.NewSource(seed+6)))
}

// gcounterLoop ticks every [1s, 5s) and alternates Increment / Get on the
// shared GCounter key. WriteLocal lets the periodic gossip loop carry the
// state to peers — this matches Scala's WriteLocal / ReadLocal usage.
func gcounterLoop(repl *ddata.Replicator, selfLabel string, rng *rand.Rand) {
	for {
		time.Sleep(ddataTickInterval(rng))
		if rng.Intn(2) == 0 {
			repl.IncrementCounter(keyGCounter, 1, ddata.WriteLocal)
		} else {
			c := repl.GCounter(keyGCounter)
			_ = c.Value()
		}
	}
}

// pncounterLoop exercises Increment / Decrement / Get with equal probability,
// mirroring the Scala-side PNCounterActor. Decrement-on-empty is a no-op in
// gekka's PNCounter implementation, so no guard is needed on cold start.
func pncounterLoop(repl *ddata.Replicator, selfLabel string, rng *rand.Rand) {
	for {
		time.Sleep(ddataTickInterval(rng))
		p := repl.PNCounter(keyPNCounter)
		switch rng.Intn(3) {
		case 0:
			p.Increment(repl.NodeID(), 1)
		case 1:
			p.Decrement(repl.NodeID(), 1)
		default:
			_ = p.Value()
		}
	}
}

// orsetLoop exercises Add / Remove / Get with equal probability over a
// 50-element key-space — matches Scala's ORSetActor.
func orsetLoop(repl *ddata.Replicator, selfLabel string, rng *rand.Rand) {
	for {
		time.Sleep(ddataTickInterval(rng))
		elem := "item-" + strconv.Itoa(rng.Intn(50))
		switch rng.Intn(3) {
		case 0:
			repl.AddToSet(keyORSet, elem, ddata.WriteLocal)
		case 1:
			repl.RemoveFromSet(keyORSet, elem, ddata.WriteLocal)
		default:
			s := repl.ORSet(keyORSet)
			_ = s.Elements()
		}
	}
}

// orflagLoop alternates SwitchOn / Get. ORFlag has no Replicator convenience
// wrapper, so we mutate the CRDT directly; gekka's gossip loop snapshots all
// known CRDTs each round and disseminates them, so the change still
// propagates on the periodic schedule. Scala's actor only ever SwitchOns
// (the Flag is monotonic), so we omit SwitchOff here too.
func orflagLoop(repl *ddata.Replicator, selfLabel string, rng *rand.Rand) {
	for {
		time.Sleep(ddataTickInterval(rng))
		f := repl.ORFlag(keyORFlag)
		if rng.Intn(2) == 0 {
			f.SwitchOn(repl.NodeID())
		} else {
			_ = f.Value()
		}
	}
}

// lwwregisterLoop alternates Set / Get. Mirrors Scala's LWWRegisterActor,
// which writes "v<0..999>" strings.
func lwwregisterLoop(repl *ddata.Replicator, selfLabel string, rng *rand.Rand) {
	for {
		time.Sleep(ddataTickInterval(rng))
		reg := repl.LWWRegister(keyLWWRegister)
		if rng.Intn(2) == 0 {
			reg.Set(repl.NodeID(), "v"+strconv.Itoa(rng.Intn(1000)))
		} else {
			_, _ = reg.Get()
		}
	}
}

// lwwmapLoop alternates Put / Get over a 10-key, 1000-value space — matches
// Scala's LWWMapActor.
func lwwmapLoop(repl *ddata.Replicator, selfLabel string, rng *rand.Rand) {
	for {
		time.Sleep(ddataTickInterval(rng))
		if rng.Intn(2) == 0 {
			k := "k" + strconv.Itoa(rng.Intn(10))
			v := "v" + strconv.Itoa(rng.Intn(1000))
			repl.PutInMap(keyLWWMap, k, v, ddata.WriteLocal)
		} else {
			m := repl.LWWMap(keyLWWMap)
			_ = m.Entries()
		}
	}
}
